internal/auth/oidc/provider.go

```diff
@@ -0,0 +1,128 @@
+package oidc
+
+import (
+	"context"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/cap/oidc"
+)
+
+var (
+	// cachedProviders provides a cache of oidc.Providers. This cache can't be
+	// done within the Repository, since a new Repository is created for every
+	// request.
+	cachedProviders     *providers
+	initCachedProviders sync.Once
+)
+
+// providerCache returns the cache of providers
+func providerCache() *providers {
+	initCachedProviders.Do(func() {
+		cachedProviders = newProviderCache()
+	})
+	return cachedProviders
+}
+
+// providers is a cache of oidc.Provider types used by the Repository to
+// make requests to the IdP, verify ID tokens, etc.  For more info on
+// oidc.Provider capabilities see: https://github.com/hashicorp/cap
+type providers struct {
+	cache map[string]*oidc.Provider
+	mu    *sync.RWMutex
+}
+
+// newProviderCache make a new cache
+func newProviderCache() *providers {
+	return &providers{
+		cache: map[string]*oidc.Provider{},
+		mu:    &sync.RWMutex{},
+	}
+}
+
+// getProvider determines if there's already a cached oidc.Provider for the
+// current AuthMethod from the DB. The cached oidc.Provider is preferred since
+// it maintains a cache of the JWKs required to verify ID Tokens from the IdP.
+//
+// Before returning a cached oidc.Provider, getProvider ensures that the
+// AuthMethod data used for the oidc.Provider's configuration hasn't changed
+// since it was cached. This is necessary because another controller could
+// update the AuthMethod in the DB, which would require changing the
+// configuration of the cached provider.
+//
+// getProvider will update the providerCache with the new AuthMethod, if it
+// determines the provider's configuration has been updated in the DB.
+func (c *providers) getProvider(ctx context.Context, currentFromDb *AuthMethod) (*oidc.Provider, error) {
+	const op = "oidc.(Repository).getProvider"
+	storedProvider, err := convertToProvider(ctx, currentFromDb)
+	if err != nil {
+		return nil, errors.Wrap(err, op)
+	}
+	c.mu.RLock()
+	p, ok := c.cache[currentFromDb.PublicId]
+	c.mu.RUnlock()
+	if ok {
+		cachedHash, err := p.ConfigHash()
+		if err != nil {
+			return nil, errors.Wrap(err, op, errors.WithMsg("unable to hash provider from cache"))
+		}
+		storedHash, err := storedProvider.ConfigHash()
+		if err != nil {
+			return nil, errors.Wrap(err, op, errors.WithMsg("unable to hash provider from database"))
+		}
+		switch cachedHash == storedHash {
```

Comment:
I love seeing the commitment to using a switch!

Reply:
I'm really trying to do this more consistently when it's an "or" condition I'm testing for.  ifs == ands and switch == ors